### PR TITLE
feat(cluster): impl slot data delete

### PR DESCRIPTION
**Pull request by adiholden**
I created FlushSlots function in db_slice. It should be called from DFLY CLUSTER CONFIG command when slots are deleted from node.

### CODE REVIEW

**Review by chakaz — APPROVED** [review]

File: src/server/db_slice.cc [hunk 1]
```diff
@@ -495,6 +495,43 @@ bool DbSlice::Del(DbIndex db_ind, PrimeIterator it) {
   return true;
 }
 
+void DbSlice::FlushSlotsFb(const absl::flat_hash_set<SlotId>& slot_ids) {
```
Comment (romange) [inline src/server/db_slice.cc:498]: If `flat_hash_set<SlotId>` is being passed a lot I recommend introducing an alias

File: src/server/db_slice.cc [hunk 2]
```diff
@@ -495,6 +495,43 @@ bool DbSlice::Del(DbIndex db_ind, PrimeIterator it) {
...
+  // Slot deletion can take time as it traverses all the database, hence it runs in fiber.
+  // We want to flush all the data of a slot that was added till the time the call to FlushSlotsFb
+  // was made. Therefore we delete slots entries with version < next_version
+  uint64_t next_version = NextVersion();
+
+  auto del_entry_cb = [&](PrimeTable::iterator it) {
+    std::string tmp;
```
Comment (romange) [inline src/server/db_slice.cc:505]: pull tmp out of the callback - to reuse it through all the iteration sequence.

File: src/server/db_slice.cc [hunk 3]
```diff
@@ -495,6 +495,44 @@ bool DbSlice::Del(DbIndex db_ind, PrimeIterator it) {
...
+    cursor = next;
+    if (i % 100 == 0) {
+      ThisFiber::Yield();
+    }
+
+  } while (cursor);
+  mi_heap_collect(ServerState::tlocal()->data_heap(), true);
+}
+
+void DbSlice::FlushSlots(const SlotSet&& slot_ids) {
```
Comment (romange) [inline src/server/db_slice.cc:529]: you do need `const` I guess

Comment (chakaz) [reply src/server/db_slice.cc:529]: Yes, a `const T&&` is something that should not exist.
If you want to take ownership over a parameter, just take it by value (i.e. `FlushSlots(SlotSet slot_ids)`). That will make it possible to call this function both with rvalues (which will be efficient and will duck an unnecessary copy), or with an lvalue (meaning just a regular variable), which will copy but will not be less efficient than accepting a `const T&`.
By accepting a `T&&` argument, you are forcing the caller to pass in a temporary, i.e. this code won't work:
```cpp
SlotSet slots;
// ...
FlushSlots(slots)
```
So only use that convention to *force* callers not to copy.

**Review by chakaz — APPROVED** [review]

**Pull request merged**